### PR TITLE
LPS-130658 | Add sample test for frontend js state

### PR DESCRIPTION
**Pull request by john-co**
**Description**
Adds a sample portlet to test the acceptance criteria for js global state epic in https://issues.liferay.com/browse/LPS-122065

**Test Plan**
1. Clicking on increment JSP button adds +1 to js state and is instantaneously reflected on JSP Counter
2. Clicking on increment react button adds +1 to js state is instantaneously reflected on the React app counter
3. Typing in the form field will change the card to show the text and string length
4. Typing in the form field will show the same text in JSP (right side) 

https://user-images.githubusercontent.com/35235266/117898116-08998d80-b279-11eb-866b-cf2d6f992e67.mp4

**JIRA Ticket**
https://issues.liferay.com/browse/LPS-130658

### CODE REVIEW

File: modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/js/sharedState.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+	return `${user.name} (${user.name.length})`;
+});
+
+export const counterAtomReact = State.atom('sample-counter', 0);
+
+// export const SharedState = {
+// 	userAtom,
+// 	userSelector,
+// }
```
Comment (kresimir-coko) [inline modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/js/sharedState.js:34]: How come we have these comments here?

File: modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/view.jsp [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+				nameElement.innerText = event.name;
+			});
+
+			buttonElementJSP.addEventListener('click', function() {
+				State.write(counterAtom, State.read(counterAtom) + 1);
+			});
+
+			buttonElementReact.addEventListener('click', function() {
+				State.write(SharedState.counterAtomReact, State.read(SharedState.counterAtomReact) + 1);
+			});
```
Comment (kresimir-coko) [inline modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/view.jsp:45]: I think we want sanity checks here if these 2 elements exist

Comment (john-co) [reply modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/view.jsp:45]: Thanks! Do you mean assert that these button elements exist in our future poshi test?

File: modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/js/sharedState.js (hunk 1 above)
Comment (john-co) [reply modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/js/sharedState.js:34]: Good point, leftover cleanup to do. I was tinkering with how to export all the variables.

**Comment by john-co** [comment]
ci:test:sf

**Comment by john-co** [comment]
ci:test:sf

**Comment by john-co** [comment]
ci:test:sf

**Comment by john-co** [comment]
ci:test:sf

**Comment by bryceosterhaus** [comment]
Hey @john-co, is the plan to start committing these test portlets to master or just to run against CI? I'm not sure if I missed a previous discussion or issue around this topic.

**Comment by john-co** [comment]
We're committing the test portlets to master in that the test app code will sit next to what it's testing. However, the test app won't be packaged with release bundle. It will only be built and deployed for testing against CI.

**Comment by john-co** [comment]
@zhangyan0701 , please forward if looks good to you. Would like your review since you've created a sample portlet before and had review feedback from BChan experience with it.

**Review by zhangyan0701 — APPROVED** [review]

**Comment by zhangyan0701** [comment]
ci:forward

**Comment by zhangyan0701** [comment]
ci:forward

File: modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/view.jsp (hunk 2 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/view.jsp]: I mean having `if (element) {}` wrapped around the usage of the `element`:

For example

```js
if (buttonElementJSP) {
	buttonElementJSP.addEventListener('click', function() {
		State.write(counterAtom, State.read(counterAtom) + 1);
	});
}
```

Comment (john-co) [reply modules/apps/frontend-js/frontend-js-state-sample-web/src/main/resources/META-INF/resources/view.jsp]: oh that makes sense, I can add those in. I'm curious, why are we more interested in sanity checking the buttonElements exist vs. the counter and name elements?

**Comment by john-co** [comment]
ci:test:relevant

**Comment by bryceosterhaus** [comment]
@john-co is the plan to continue adding test portlets for features? And if there has already been a long previous conversation about this, forgive me for bringing it up again 😂 . 

My hesitation is that we start to have an additional QA module for every module that gets bundled. One thought would be have single QA-type module at the root of our modules. So something like `frontend-js/frontend-js-qa`, `frontend-taglib/frontend-taglib-qa`, and then within those modules you could have multiple types of tests. So the js-qa portlet would contain ways to test multiple modules rather than just one. Am I way off in my thought process here, I may be overthinking it.

**Comment by john-co** [comment]
Hi @bryceosterhaus , yea I was thinking about that too and how it could create a lot of clutter with multiple qa modules. I think that's a good idea to have a general qa module that could test all the other sub-modules. I'll update the name and we'll go with that direction.

**Comment by bryceosterhaus** [comment]
@john-co sounds good! And if you need any help setting up some tabs or pages so that you can separate some tests, let me know and I can send something your way

**Comment by john-co** [comment]
Going off of BChan's comment here: https://github.com/brianchandotcom/liferay-portal/pull/100733#issuecomment-818674441

We'll probably want to name it something like `frontend-js-sample-web`

**Comment by john-co** [comment]
ci:test:sf

**Comment by john-co** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:forward

**Pull request closed without merging**